fix: getSchemaSerializer contentType check

salesh opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->
https://github.com/fastify/fastify/issues/4528

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

climba03003 reviewed (changes requested):
Can you add a unit test to prevent the same problem happen again?

salesh commented:
I would need a little help with that 

I had the idea to expand  https://github.com/fastify/fastify/blob/a71dd83cc65c79c2a01dfbe91f8ee64fec3c0509/test/hooks.test.js#L1271  with

```
  fastify.post('/', {
    schema: {
      response: {
        500: {
          content: {
            'application/json': {
              schema: {
                name: { type: 'string' },
                image: { type: 'string' },
                address: { type: 'string' }
              }
            }
          }
        }
      }
    }
  }, (req, reply) => {
    reply.send({ hello: 'world' })
  })
```
I tried to make the test fail for code state before this PR and I added some logs to observe it

```
function getSchemaSerializer (context, statusCode, contentType) {
  const responseSchemaDef = context[kSchemaResponse]
  if (!responseSchemaDef) {
    return false
  }
  if (responseSchemaDef[statusCode]) {
    console.log('TEST1', responseSchemaDef[statusCode].constructor === Object, contentType)
    if (responseSchemaDef[statusCode].constructor === Object) {
      console.log('TEST2', contentType, typeof contentType)
      const mediaName = contentType.split(';')[0]
      console.log('TEST3', contentType.split(';')[0])
      if (responseSchemaDef[statusCode][mediaName]) {
        return responseSchemaDef[statusCode][mediaName]
      }

      return false
    }
    return responseSchemaDef[statusCode]
  }
```

```
    Before fix
    
    # Subtest: onSend hook throws
        1..11
        ok 1 - should not error
    TEST1 true undefined
    TEST2 undefined undefined
    TEST1 true undefined
    TEST2 undefined undefined
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
    fallbackErrorHandlerfallbackError 500 application/json; charset=utf-8
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
    fallbackErrorHandlerfallbackError 500 application/json; charset=utf-8
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
    
    You can see that there is no TEST3 -> it seems like he somehow swallow it or he recreates another error that got covered (that's why he is not reporting and test pass)
    
    After fix
    
	# Subtest: onSend hook throws
    	1..11
   	 ok 1 - should not error
    TEST1 true undefined
    TEST1 true application/json; charset=utf-8
    TEST2 application/json; charset=utf-8 string
    TEST3 application/json
    fallbackErrorHandlerfallbackError 500 application/json; charset=utf-8
    packet {"name":"Error"}
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
    fallbackErrorHandlerfallbackError 500 application/json; charset=utf-8
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
    fallbackErrorHandlerfallbackError 500 application/json; charset=utf-8
    packet {"statusCode":500,"error":"Internal Server Error","message":"some error"}
```


So is it better to test individual function? If yes can you suggest to me the place where I should put it. @climba03003

climba03003 reviewed (commented):
You can mock the `getSchemaSerializer` function and assert if the last parameter is passed in or not.

mcollina reviewed (changes requested):
Thanks for opening a PR! Can you please add a unit test?

salesh commented:
Sorry for my slowness guys, I am not the best with a tap.



> You can mock the `getSchemaSerializer` function and assert if the last parameter is passed in or not.

Are you suggesting that I should add a mock inside this test that I mention in https://github.com/fastify/fastify/pull/4531#issuecomment-1403409944 ?

Would you mind helping me to understand where am I making mistake in

```
  t.mock('../lib/schemas', {
    getSchemaSerializer: (param1, param2, param3) => {
      t.equal(param3, 'application/json; charset=utf-8', 'param3 should be "application/json; charset=utf-8"')
      t.end()
    }
  })
```
I keep getting 

```
 ✖ Cannot find module 'getSchemaSerializer' Require stack: - /home/x/x/x/fastify/test/hooks.test.js

  test: onSend hook throws
  stack: |
    Test.<anonymous> (test/hooks.test.js:1291:5)
  at:
    line: 933
    column: 15
    file: node:internal/modules/cjs/loader
    function: Function.Module._resolveFilename
  code: MODULE_NOT_FOUND
  requireStack:
    - /home/x/x/x/fastify/test/hooks.test.js
  tapCaught: testFunctionThrow
```

@climba03003

climba03003 commented:
Should be something like this.
```js
const Fastify = t.mock('../', {
  '../lib/schemas': {
    getSchemaSerializer: (param1, param2, param3) => {
      t.equal(param3, 'application/json; charset=utf-8', 'param3 should be "application/json; charset=utf-8"')
      t.end()
    }
  }
})
```

salesh commented on test/hooks.test.js:
```diff
@@ -1269,8 +1269,16 @@ test('clear payload', t => {
+        t.end()
+      }
+    }
+  })
```
I didn't know that this needs to be done like this, thank you @climba03003 


I am now trying to figure out why 
```js
 FAIL ​ test/hooks.test.js
 ✖ this.opts.bucket is not a function

    } else {
      this.schemaBucket = this.opts.bucket()
------------------------------------^
    }
  }

  test: onSend hook throws
  stack: |
    new SchemaController (lib/schema-controller.js:55:37)
    Function.buildSchemaController (lib/schema-controller.js:37:10)
    fastify (fastify.js:206:45)
    Test.<anonymous> (test/hooks.test.js:1281:19)
  at:
    line: 55
    column: 37
    file: lib/schema-controller.js
    function: SchemaController
  type: TypeError
  tapCaught: testFunctionThrow
```
is happening.

climba03003 replied:
Maybe it would be easier to use `proxyquire`. It can replace the necessary only.

```js
const proxyquire = require('proxyquire')
const Fastify = proxyquire('..', {
  '../lib/schemas': {
    getSchemaSerializer: (param1, param2, param3) => {
      t.equal(param3, 'application/json; charset=utf-8', 'param3 should be "application/json; charset=utf-8"')
      t.end()
    }
  }
})
```

climba03003 commented on test/hooks.test.js:
```diff
@@ -1269,7 +1270,15 @@ test('clear payload', t => {
+    './lib/schemas.js': {
+      getSchemaSerializer: (param1, param2, param3) => {
+        t.equal(param3, 'application/json; charset=utf-8', 'param3 should be "application/json; charset=utf-8"')
+        t.end()
```
Do not end the test yourself.
```suggestion
```

salesh replied:
> Do not end the test yourself.

Done. I would appreciate it if you could share it with me why?

climba03003 replied:
t.plan will assert if the assertion number is match what it's needed and end accordingly.

When you forcefully end the test. All assertion check will not run at that point and end no matter it success or fail.

salesh replied:
> t.plan will assert if the assertion number is match what it's needed and end accordingly.
> 
> When you forcefully end the test. All assertion check will not run at that point and end no matter it success or fail.

Thank you for sharing your knowledge with me and for your time @climba03003

mcollina reviewed (approved):
lgtm
